Cookbook/Technical-Documents/From_0%_to_100%_with_Bento.md

```diff
@@ -0,0 +1,443 @@
+
+```plain
+Bento now is our internal library however we used to open source it. It has changed when Apple announced SwiftUI at WWDC 2019. We decided to move it under the main repo back then.
+```
+
+## This is what we want to build
+![](/Users/adam.borek/Developer/babylon/images/bento01.png)
+
+## Renderer
+I assume you are already kind of familiar with our Architecture. Quick recap, a typical screen is built from:
+- ViewController
+- ViewModel
+- Renderer
+- Builder
+- FlowController
+
+In this article we are going to focus on the renderer. 
+
+Renderer purpose is to calculate a UI. I used the word calculate on purpose as we express our UI as a function of state `UI = f(state)`. The function is named `func render(state: ViewModel.State) -> Screen<SectionId, ItemId>`.
+
+First things first. To create a Renderer you need to create a struct and conforms to `BoxRenderer`. This is what you get from the Xcode's template.
+
+```swift
+struct ChoosePharmacyTypeRenderer: BoxRenderer {
+    private let config: Config
+    private let observer: Sink<ChoosePharmacyTypeViewModel.Action>
+
+    init(
+        observer: @escaping Sink<ChoosePharmacyTypeViewModel.Action>,
+        appearance: BabylonAppAppearance,
+        config: Config
+    ) {
+        self.config = config
+        self.observer = observer
+    }
+
+    func render(state: ChoosePharmacyTypeViewModel.State) -> Screen<SectionID, NodeID> {
+        return Screen(title: "", box: .empty)
+    }
+
+    struct Config {
+        let bundle: Bundle
+
+        init(bundle: Bundle = .main) {
+            self.bundle = bundle
+        }
+    }
+}
+
+extension ChoosePharmacyTypeRenderer {
+    enum SectionID: Hashable {
+        case first
+    }
+
+    enum NodeID: Hashable {
+        case first
+    }
+}
+
+```
+
+Going from the top the first question you may ask is "What's a Config". As a result of how `BoxRenderer` is coupled with `BabylonBoxViewController` we cannot pass dependencies as we usually do in a init. Config is a way of injecting dependencies into a renderer.
+
+```plain
+ 12 Aug 2019 - Renderer has to be a struct as there is a problem with classes & memory management. Classes are not deallocated. It may be fixed someday but it's still the case when I'm writing this article.
+```
+
+Now let's jump into the `render(state:)`. It returns a Screen. This is how a Screen may look like:
+```swift
+    return Screen(
+	    title: screenTitle,
+	    rightBarItems: [faqButton],
+	    shouldUseSystemSeparators: false,
+	    box: render(state),
+	    pinnedToBottomBox: renderBottomButtons(state)
+    )
+```
+
+With Screen you can modify many behaviour like alignment of the content on the screen, whether you want system's separators or not. If you need to modify something which cannot be represented by a box please take a look on the init of the `Screen`. Variables there are pretty self explanatory.
+
+#### What's the Box then?
```

Comment:
I'm not sure if it's worth linking to this for further reading https://github.com/Babylonpartners/Bento#how-does-it-work-

In fact - I think a link to the Bento repo should be in this article somewhere anyway

Reply:
I don't know if it's good to link to a repo which is no longer maintained. The repo still uses Stylesheets while we have moved to operators